Properly support updating dependencies

bjgill opened the pull request:
Now, if we attempt to add a dependency that already exists, we will not
throw away information. i.e. an optional dependency will always remain
as such.

Also some minor refactoring to use PathBuf instead of String where
appropriate.

This is a necessary prerequisite for a follow-on PR that implements a `cargo upgrade`

bjgill commented:
OK, the follow-on that implements `cargo upgrade` is ready (https://github.com/bjgill/cargo-edit/tree/cargo-upgrade). I'll submit once this is merged.

killercup commented:
As mentioned in <https://github.com/killercup/cargo-edit/pull/132#issuecomment-312633079> I'll not have time review much this week. So,

r? @ordian

ordian commented:
hi @bjgill! 
Currently ```cargo add foo --path=bar``` will append the following table
```
[dependency.foo]
optional = false
path = "bar"
```
to Cargo.toml (same for --git dependency), which is weird by itself.
But after this change, ```cardo add foo --path=bar``` followed by ```cargo add foo --git=baz``` will create a mess
```
[dependency.foo]
git = "baz"
optional = false
path = "bar"
```

Also it would be nice to specify (in usage of --update-only) that it will only update the version (and won't touch e.g. features). 

FYI, I'm working on toml_document's [fork](https://github.com/ordian/toml_document), so maybe we will switch to that instead of toml crate.

bjgill commented:
Good point. We only really need this behaviour for `cargo add --update-only`, so I'll switch the merging to only happen on update.

ordian commented:
How about ```cardo add foo --path=bar``` followed by ```cargo add foo --git=baz --update-only```?
Why do we need merge if we only need to update the version?

bjgill commented:
OK - I think that's fixed now, though it would be nice to have tests to confirm.

I think the next steps are:

- Move plain `cargo add` back to merging behaviour. Now that merging works correctly, I see no reason not to.
- Add tests to verify that `cargo add` works correctly when used on existing dependencies.
- Remove `cargo add --update-only` and replace with `cargo upgrade` (in follow-up PR)

I'll look to trying to get that sorted out tomorrow.

bjgill commented:
OK - I've switched `cargo add` back over to merging behaviour, and added some tests (which caught an edge-case I'd missed).

@ordian - If you're happy, I'll squash and merge this, then produce the follow-on PR which implements a proper `cargo upgrade`

ordian commented on src/dependency.rs:
```diff
@@ -93,7 +93,9 @@ impl Dependency {
                     }
                 }
-                data.insert("optional".into(), toml::Value::Boolean(optional));
+                if self.optional {
```
This is nice.

ordian commented:
LGTM, thanks!

bjgill commented:
bors r+
